### PR TITLE
SocketCAN

### PR DESCRIPTION
**Pull request by andreeatulbure**
SocketCAN implementation under the supervision of Denis Stogl

### CODE REVIEW

File: schunk_libm5api/src/Device/Device.cpp [hunk 1]
```diff
@@ -4805,21 +4806,26 @@ CDevice* newDevice(const char* acInitString)
...
+#endif
 #ifdef USE_PCAN
 	if( strcmp( pcToken, "PCAN" ) == 0 )
 	{
 		return new CPCanDevice();
 	}
 #endif
-#ifdef USE_ESD
-	if( strcmp( pcToken, "ESD" ) == 0 )
+#ifdef USE_SCAN
```
Comment (fmessmer) [inline schunk_libm5api/src/Device/Device.cpp]: I think `SCAN` might be miss-leading....wouldn't `USE_SOCKET_CAN` be clearer?

File: schunk_libm5api/src/M5apiw32/m5apiw32.cpp [hunk 2]
```diff
@@ -1230,7 +1230,7 @@ M5DLL_API int WINAPI PCube_getStateDioPos( int iDeviceId, int iModuleId, unsigne
 		return ERRID_DEV_NOTINITIALIZED;
 
 	int iRetVal = g_apclDevice[iDeviceId]->getStateDioPos( iModuleId, puiState, pucDio, pfPos);
-
+
```
Comment (fmessmer) [inline schunk_libm5api/src/M5apiw32/m5apiw32.cpp]: undo trailing white spaces

File: schunk_libm5api/src/Device/SocketCANDevice.h [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+#define AF_CAN PF_CAN
+#endif
+
+
+//-----------------------------------------------
+
+#include <iostream>
+#include <cstdio>
+#include <libpcan/libpcan.h>
+#include "../Device/ProtocolDevice.h"
```
Comment (fmessmer) [inline schunk_libm5api/src/Device/SocketCANDevice.h]: use proper include?...no relative paths

File: schunk_libm5api/src/Device/Device.cpp (hunk 1 above)
Comment (andreeatulbure) [inline schunk_libm5api/src/Device/Device.cpp]: I first tried to use USE_SOCKET_CAN but it was to long and it didn't work. This is why I shortened it and wrote USE_SCAN.

**Comment by fmessmer** [comment]
@ipa-mdl would you like to comment on this PR?

Unfortunately, I cannot give qualified comments about implementation and functionality...

Still, I think for this PR @ipa-mdl's [comment](https://github.com/ipa320/cob_driver/pull/269#issuecomment-198391270) applies, too.
Thus, we should hold this PR and merge together with https://github.com/ipa320/cob_driver/pull/269 (?)

**Comment by destogl** [comment]
@ipa-fxm, regarding @ipa-mdl's [comment](https://github.com/ipa320/cob_driver/pull/269#issuecomment-198391270): in this case it doesn't apply since implementation here is not depending on [ ros-industrial/ros_canopen](https://github.com/ros-industrial/ros_canopen) package. Here we have direct implementation of socket_can interface. We decided to use it, since we had this implementation already.

File: schunk_libm5api/src/Device/SocketCANDevice.h [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <sys/socket.h>
+#include <sys/ioctl.h>
+
+#include <fcntl.h>
+#include <cerrno>
+#include <cstring>
+
+#include <linux/can.h>
```
Comment (mathias-luedtke) [inline schunk_libm5api/src/Device/SocketCANDevice.h:55]: This needs the `linux-headers-generic` dependency.

**Comment by mathias-luedtke** [comment]
@destogl is right, no need to couple it with the `cob_generic_can` PR.

File: schunk_libm5api/src/Device/Device.cpp (hunk 1 above)
Comment (mathias-luedtke) [inline schunk_libm5api/src/Device/Device.cpp]: `USE_SOCKET_CAN"  should be used.

File: schunk_libm5api/src/Device/SocketCANDevice.h (hunk 4 above)
Comment (destogl) [inline schunk_libm5api/src/Device/SocketCANDevice.h:55]: Done!

**Comment by andreeatulbure** [comment]
When are you planning to merge this?

**Comment by mathias-luedtke** [comment]
At least the `USE_SOCKET_CAN` issue should be resolved.

**Comment by destogl** [comment]
I can check this, but it is problem by variable definition. We give proposal for solution. @andreeatulbure can you look at this?

**Comment by andreeatulbure** [comment]
We did the changes. It should be ok now.

**Comment by mathias-luedtke** [comment]
It looks okay now.
Before the merge it needs to be tested with SocketCAN and without.
I don't have any powercube devices.

Can you test both, please?

**Comment by destogl** [comment]
Everything is tested and works! So you can merge.

File: schunk_libm5api/src/Device/Device.cpp [hunk 5]
```diff
@@ -4823,7 +4823,7 @@ CDevice* newDevice(const char* acInitString)
 	}
 #endif
 #ifdef USE_SOCKET_CAN
-        if( strcmp( pcToken, "SCAN" ) == 0 )
+        if( strcmp( pcToken, "SOCEKTCAN" ) == 0 )
```
Comment (ipa-mdl) [inline schunk_libm5api/src/Device/Device.cpp]: Typo

**Comment by mathias-luedtke** [comment]
+1 for merge

**Comment by NadiaHG** [comment]
FR: OK

**Pull request merged**